_posts/2019-09-10-unbearable-lightness-of-refactoring.markdown

```diff
@@ -0,0 +1,115 @@
+
+### Lets Get More Specific
+
+Back to our SCA change, we ended up having some discussions about how to improve our already complicated logic in a way that easily supports future updates. We started by trying to separate the different steps that an order has to go through to fully get submitted and try to simplify it, weighing the pros and cons of each approach. While pretty much all of our solutions would have some cons, we managed to find our best option and started refactoring. Without getting too technical, lets look at the actual change.
+
+```ruby
+## before refactor
+order.submit! do
+  order.line_items.each { |li| li.update!(commission_fee_cents: li.current_commission_fee_cents) }
+  totals = BuyOrderTotals.new(order)
+  order.update!(
+   # set totals
+  )
+  order_processor.hold!
+  raise Errors::InsufficientInventoryError if order_processor.failed_inventory?
+  # in case of failed transaction, we need to rollback this block,
+  # but still need to add transaction, so we raise an ActiveRecord::Rollback
+  raise ActiveRecord::Rollback if order_processor.failed_payment? || order_processor.requires_action?
+
+  order.update!(
+    # set payment
+  )
+  order.transactions << order_processor.transaction
+  PostTransactionNotificationJob.perform_later(order_processor.transaction.id, user_id)
+  raise Errors::FailedTransactionError.new(:charge_authorization_failed, order_processor.transaction) if order_processor.failed_payment?
+  if order_processor.requires_action?
+    # because of an issue with `ActiveRecord::Rollback` we have to force a reload here
+    # rollback does not clean the model and calling update on it will raise error
+    order.reload.update!(external_charge_id: order_processor.transaction.external_id)
+    Exchange.dogstatsd.increment '******'
+    raise Errors::PaymentRequiresActionError, order_processor.action_data
+  end
+end
+```
+
+In original solution we wrapped all of our changes in a database transaction within `order.submit!` which would have done a lock on that record. This was all good since we would ensure data integrity provided by transaction where we want to make sure updates to `order` and `line_items` happen only in case of success. A failure in this block would rollback all changes which is good 👍
+
+But things got complicated once we started have situation that some of the changes during the block should have been preserved even in case of rollback. Specifically we want to make sure a `transaction` is stored on the `order` if it failed payment or requires action.
```

Comment:
```suggestion
But things got complicated once some of the changes in the block _should_ have been preserved, even in case of rollback. Specifically we want to make sure a `transaction` is stored on the `order` if it payment fails or requires action.
```